### PR TITLE
New: add callback-return rule (fixes #994)

### LINKED ISSUE #994: Rule: async callback should be paired with return
In `node.js` typical error is writing something like this:

```javascript
if (err) {
  callback(err);
}
```

Right cases are:

1.

```javascript
if (err) {
  callback(err);
  return;
}
```

2.

```javascript
if (err) return callback (err);
```

Imho, this can be automated. Preliminary conditions to detect callback

- keywords `callback`, `cb`, `next` in function name
- 0-2 params

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2728300-rule-async-callback-should-be-paired-with-return?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by xjamundx**

### CODE REVIEW

**Comment by xjamundx** [comment]
Need some feedback on the implementation of the `IfStatement` algorithm....I'm feeling like it's a bit hacky. Is there some sort of `context.getNextNode()`?

File: lib/rules/callback-return.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+                // if the next thing that occurs is a return statement, we"re happy
+                if (keywords[0] !== "return") {
+                    context.report(node, "Expected return with your callback function.");
+                }
+
+            }
+
+            // if the parent is a function then we can do whatever
+            // TODO: arrow functions?
```
Comment (xjamundx) [inline lib/rules/callback-return.js]: Actually I don't think I need this whole block...

File: lib/rules/callback-return.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            if (isCallback(node)) {
+                parent = findClosestParentOfType(node, ["ReturnStatement", "FunctionExpression", "FunctionDeclaration", "IfStatement"]);
+            }
+
+            if (!parent) {
+                return;
+            }
+
+            // if our parent is a return statement we should be happy
+            if (parent.type === "ReturnStatement") {
```
Comment (xjamundx) [inline lib/rules/callback-return.js]: Not even needed.

File: lib/rules/callback-return.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            if (parent.type === "FunctionExpression" || parent.type === "FunctionExpression") {
+                return;
+            }
+
+        }
+
+    };
+};
+
+module.exports.schema = [];
```
Comment (xjamundx) [inline lib/rules/callback-return.js]: Oh I need a schema

File: docs/rules/callback-return.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
+# Enforce Semicolons (callback-return)
```
Comment (mathieumg) [inline docs/rules/callback-return.md]: Wrong header name?

Comment (xjamundx) [inline docs/rules/callback-return.md]: haha apparently

File: conf/environments.js [hunk 5]
```diff
@@ -25,7 +25,8 @@ module.exports = {
             "no-path-concat": 2,
             "no-process-exit": 2,
             "global-strict": [0, "always"],
-            "handle-callback-err": [2, "err"]
+            "handle-callback-err": [2, "err"],
+            "callback-return": 0
         }
```
Comment (nzakas) [inline conf/environments.js]: This isn't necessary, it's already off by default.

File: docs/rules/callback-return.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
+# Enforce Return After Callback (callback-return)
+
+The callback pattern is at the heart most I/O and event programming
+ in javascript.
+
```
Comment (nzakas) [inline docs/rules/callback-return.md]: JavaScript :)

File: docs/rules/callback-return.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+```json
+callback-return: [2, ["callback", "cb", "next"]]
+```
+
+### Gotchas
+
+There are many cases of bad behavior that this rule will not catch and even a few cases where you're doing the right
+thing but this rule will warn, primarily because of the difficulty of static analyzing intent.
+
+###### Passing the Callback by Reference
```
Comment (nzakas) [inline docs/rules/callback-return.md]: A few too many pound signs

File: lib/rules/callback-return.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                parent = findClosestParentOfType(node, ["ReturnStatement", "FunctionExpression", "FunctionDeclaration", "IfStatement"]);
+            }
+
+            if (!parent) {
+                return;
+            }
+
+            // if you we are in an if statement we need to ensure we return soon
+            if (parent.type === "IfStatement") {
+
```
Comment (nzakas) [inline lib/rules/callback-return.js]: Why is an if statement special?

**Comment by nzakas** [comment]
I'm not sure about the approach, it seems fragile to use n number of tokens after a call to determine if a return is present. Can you use the traversal to help?

Also, can you turn this on for the project? I want to get back to having as many rules on for this project as possible (dogfood).

**Comment by xjamundx** [comment]
Fixed the little things, going to think about a better system for determining failure.....agreed it's brittle!

**Comment by xjamundx** [comment]
Need to also support things like `callback && callback()` and possibly even `if (callback) callback();'

**Comment by xjamundx** [comment]
Current logic allows for this. Will fix and add test.

```js
function a() {
   callback();
   callback();
}
```

**Comment by xjamundx** [comment]
Stealing this from `no-unreachable` 

```js
    /*
     * Verifies that the given node is the last node or followed exclusively by
     * hoisted declarations
     * @param {ASTNode} node Node that should be the last node
     * @returns {void}
     * @private
     */
    function checkNode(node) {
```

File: lib/rules/callback-return.js [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                // try to figure out the next thing that happens
+                nextNodes = context.getTokensAfter(node, 10); // punctuation, comment, else, return, whatever
```
Comment (lo1tuma) [inline lib/rules/callback-return.js]: Why exactly 10 tokens? This seems completely arbitrary.

This would fail:
```js
callback();;;;;;;;;;;
return;
```

Or this:
```js
((((((((((callback()))))))))))
return;
```

Comment (xjamundx) [inline lib/rules/callback-return.js]: haha yes of course it will fail. looking for suggestions on another way. currently messing with some of the code from `no-unreachable`  to find something more robust. so far no luck, but not giving up.

Comment (ilyavolodin) [inline lib/rules/callback-return.js]: It feels like, again, we need shared code for checking all branches of AST... This is like the 10th time we ran into this.

**Comment by lo1tuma** [comment]
Another approach would be to use escope for this:

1. Listen to `FunctionDeclaration`, `FunctionExpression` and `ArrowFunctionExpression`
1. get the current scope
1. get all variables of type `Parameter` from the current scope
1. find all callback variables which are matching one of the configured keywords
1. validate all references of the found callback variables to be paired with an return. For each reference:
  1. check if the reference scope (`reference.from`) is the same as the current scope
  1. check if the parent node of the reference is a `CallExpression` (you need to use `context.getNodeByRangeIndex` for this because escope nodes doesn’t have a `parent` property)
  1. check if the `CallExpression` is paired with a `ReturnStatement`

File: lib/rules/callback-return.js (hunk 9 above)
Comment (lo1tuma) [inline lib/rules/callback-return.js]: @ilyavolodin this is not about branch tracking, it is about getting the next AST node.

@xjamundx Have you tried `context. getNodeByRangeIndex(node.range[1] + 1);`?

**Comment by xjamundx** [comment]
Trying a different route which involves lots of going up and down the various trees. I'm worried my approach my be missing some important cases or fail in some areas. So far so good though! 32 passing tests

File: lib/rules/callback-return.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        return node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1;
+    }
+
+    /**
+     * Determines whether or not the callback node is classified as a callback.
+     * @param {ASTNode} node The callback node
+     * @param {ASTNode} parentNode The parent node
+     * @returns {boolean} Whether or not this is part of a callback expression
+     */
+    function isCallbackExpression(node, parentNode) {
```
Comment (xjamundx) [inline lib/rules/callback-return.js]: I may be able to merge this with the `isCallback` function...and simplify some of the code...trying.

File: lib/rules/callback-return.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+     * Check to see if a CallExpression is in our callback list.
+     * @param {ASTNode} node The node to check against our callback names list.
+     * @returns {Boolean} Whether or not this function matches our callback name.
+     */
+    function isCallback(node) {
+        return node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1;
+    }
+
+    /**
+     * Determines whether or not the callback node is classified as a callback.
```
Comment (xjamundx) [inline lib/rules/callback-return.js]: minor update coming to the explanation here

**Comment by xjamundx** [comment]
Let me know if there are any further improvements that need to be made (or possible test cases)

**Comment by xjamundx** [comment]
Haven't considered class methods. Let me know if we need to add tests for them in the case that they deal with callbacks.

**Comment by nzakas** [comment]
Yup, we should cover class methods

**Comment by xjamundx** [comment]
Okay, just needed some tests. It worked.

**Comment by nzakas** [comment]
Looks like Travis is failing due to a code coverage issue

**Comment by xjamundx** [comment]
100% coverage!

File: docs/rules/callback-return.md [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+The rule takes a single option, which is an array of possible callback names.
+
+```json
+callback-return: [2, ["callback", "cb", "next"]]
+```
+
+### Gotchas
+
+There are many cases of bad behavior that this rule will not catch and even a few cases where you're doing the right
+thing but this rule will warn, primarily because of the difficulty of static analyzing intent.
```
Comment (ilyavolodin) [inline docs/rules/callback-return.md]: The ending is hard to understand. Maybe "difficulty of determining intent through static code analysis"?

**Comment by xjamundx** [comment]
Changed the text in question @ilyavolodin . Hope that's clearer now! Thanks for catching.

**Comment by nzakas** [comment]
Lgtm. @ilyavolodin ?

File: lib/rules/callback-return.js [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    return {
+        "CallExpression": function(node) {
+
+            // if we"re not a callback we can return
+            if (!isCallback(node)) {
+                return;
+            }
+
+            // find the closest block, return or if statement
+            var closestBlock = findClosestParentOfType(node, ["BlockStatement", "ReturnStatement", "IfStatement"]),
```
Comment (lo1tuma) [inline lib/rules/callback-return.js]: What about `SwitchCase`?

Comment (xjamundx) [inline lib/rules/callback-return.js]: Can you think of a quick case for `switch` and I'll add it?

Comment (ilyavolodin) [inline lib/rules/callback-return.js]:
```js
function a(value, callback) {
  switch {
    case 1: callback(22); break;
    case 2: callback(33); break;
  }
  callback(0);
}
```
Would that work?

Comment (lo1tuma) [inline lib/rules/callback-return.js]: I’m not sure if this special treatment (only) for if statements is a good idea at all. If we go this way, we should also include loops.

After reading through the issue again, I also can’t find that this was requested.

So I would actually expect this rule to **always** warn if a callback call is not paired with a return statement, event the following super simple case:

```js
function foo(callback) {}
    callback(); // <-- should be a warning
}
```

Comment (xjamundx) [inline lib/rules/callback-return.js]: You specifically talking about line 107? https://github.com/eslint/eslint/pull/2828/files#diff-c53c86e3156c85a799267c9ec841d054R107

File: lib/rules/callback-return.js [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            // block statements are part of functions and most if statements
+            if (closestBlock.type === "BlockStatement") {
+
+                // find the last item in the block
+                lastItem = closestBlock.body[closestBlock.body.length - 1];
+
+                // if the callback is the last thing in a block that might be ok
+                if (isCallbackExpression(node, lastItem)) {
```
Comment (lo1tuma) [inline lib/rules/callback-return.js]: This would not catch the following example:

```js
function foo(callback) {
    {
        callback();
    }
    
    callback();
}
```

Comment (xjamundx) [inline lib/rules/callback-return.js]: I think most of the real-world cases we were hoping to catch were inside of `IfStatements`. I'm willing to remove that though and just say you *always* have to return with a callback everywhere. Thoughts?

File: lib/rules/callback-return.js [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+                    // but only if the callback is immediately before
+                    if (isCallbackExpression(node, closestBlock.body[closestBlock.body.length - 2])) {
+                        return;
+                    }
+                }
+
+            }
+
+            // at this point we're likely doing something like callback(); callback();
```
Comment (lo1tuma) [inline lib/rules/callback-return.js]: Can you add a test `callback(); callback();`?

Comment (xjamundx) [inline lib/rules/callback-return.js]: We have that inside of a function. you mean outside of one?

File: lib/rules/callback-return.js (hunk 13 above)
Comment (lo1tuma) [inline lib/rules/callback-return.js]: Yes.

**Comment by xjamundx** [comment]
I feel weird demanding people `return` when they're already at the end of the function. My preference is to limit this to `IfStatement` and `SwitchCase` blocks. Happy to be wrong....

File: lib/rules/callback-return.js (hunk 15 above)
Comment (lo1tuma) [inline lib/rules/callback-return.js]: You are right, sorry I didn’t see it at the first glance.

**Comment by lo1tuma** [comment]
For me it feels like a inconsistent behavior, because you would allow callbacks to be called without return in loops.

File: lib/rules/callback-return.js (hunk 14 above)
Comment (lo1tuma) [inline lib/rules/callback-return.js]: Here is another, more "real-world", example:

```js
function foo(callback) {
    try {
        a();
        callback();
    } catch (err) {}
    
    try {
        b();
        callback();
    } catch (err) {}
}
```

**Comment by xjamundx** [comment]
So how about this:

1. `callback()` must be preceded by or followed by a `return` when used in a block
2. `callback()` (or `return`) must also be the last item in the block
2. In a function block `callback()` assuming 2 you don't need to `return`

**Comment by mathieumg** [comment]
> I feel weird demanding people return when they're already at the end of the function.

I feel the same. Many times the last item in the block is a call to something which expects a callback as one of the arguments, I imagine that call itself could be returned but it feels weird to have to do that as it's the last line.

**Comment by lo1tuma** [comment]
@xjamundx I would rephrase it a little bit:

`callback()` must be preceded by or followed by a `return`.

Exceptions:
1. The direct parent is a function block AND the `callback()` is the last statement inside the block (or 2nd last if `return` is last)
2. The direct parent is a logical expression in form of `callback && callback()` AND the grand parent is a function block AND the logical expression is the last statement inside the block (or 2nd last if `return` is last)

**Comment by xjamundx** [comment]
@lo1tuma thanks for your extremely detailed feedback. I've now updated the PR to account for the latest best understanding of how this should work, adding test for loops and other things without the rule explicitly knowing about them.

...It's amazing how many different implementations can lead to passing tests. Hope I finally go it!

File: tests/lib/rules/callback-return.js [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+        "function x(err) { if (err) { \n log();\n callback(); return; } }",
+        "function x(err) { if (err) { callback(); return; } return callback(); }",
+        "function x(err) { if (err) { return callback(); } else { return callback(); } }",
+        "function x(err) { if (err) { return callback(); } else if (x) { return callback(); } }",
+        "function x(err) { if (err) return callback(); else return callback(); }",
+        "function x(cb) { cb && cb(); }",
+        "function x(next) { typeof next !== 'undefined' && next(); }",
+        "function x(next) { if (typeof next === 'function')  { return next() } }",
+        "function x() { switch(x) { case 'a': return next(); } }",
+        "function x() { switch(x) { case 'a': return next(); } }",
```
Comment (xjamundx) [inline tests/lib/rules/callback-return.js]: dup. let me switch to `for` loop....

**Comment by xjamundx** [comment]
Realized we shouldn't ask people to `return` unless they're inside of a function. 1 more pass...

**Comment by lo1tuma** [comment]
LGTM :+1:

**Comment by ilyavolodin** [comment]
LGTM, will hold off on merging this till we sure we don't need patch releases (and I need to figure out where to merge this into, master or 0.x branch)

**Comment by nzakas** [comment]
Master. We will only make changes to 0.x for important bug fixes that can't wait for 1.0.0
